### PR TITLE
Support HDFS Artifact

### PR DESCRIPTION
**Pull request by dtaniwaki**
I implemented the HDFS artifact suggested in https://github.com/argoproj/argo/issues/1140.

I have tested it with [kubernetes-HDFS](https://github.com/apache-spark-on-k8s/kubernetes-HDFS) and confirmed it works with no-auth, auth by ccache, auth by keytab with the patch mentioned in https://github.com/argoproj/argo/issues/1140#issuecomment-451628373.

Here're several items to discuss.

1. I used `Addreseses` for the config because the client only needs the information, but just loading XML files of `hdfs-site.xml` from a configmap like `KrbConfigConfigMap` may be easier for users although it contains lots of unnecessary information.
2. I made some refactoring in the existing methods to put the code to create an HDFS driver and validate an HDFS artifact in the hdfs package. I'm not sure if this refactoring is also good for other artifacts.
3. I used the revision `48eb8d6c34a97ffc73b406356f0f2e1c569b42a5` of  github.com/colinmarc/hdfs instead of the latest `v2.0.0` because `v2.0.0` is structured as a go module but I can't make it work in the argo repo.
4. There's a limitation of HDFS setting because [the HDFS client](github.com/colinmarc/hdfs) of `48eb8d6c34a97ffc73b406356f0f2e1c569b42a5` (also `v2.0.0`) doesn't work with `hadoop.rpc.protection=privacy`. Is it acceptable?

### CODE REVIEW

File: workflow/artifacts/hdfs/hdfs.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	if art.KrbCCacheSecret != nil && art.KrbCCacheSecret.Name != "" {
+		bytes, err := ci.GetSecrets(namespace, art.KrbCCacheSecret.Name, art.KrbCCacheSecret.Key)
+		if err != nil {
+			return nil, err
+		}
+		ccache, err := credentials.ParseCCache(bytes)
+		if err != nil {
+			return nil, err
+		}
+		fmt.Printf("ccache: %v", ccache)
```
[the remark this answers is not in the document]
Comment (dtaniwaki) [reply workflow/artifacts/hdfs/hdfs.go]: Oops, it’s a log for test. I’ll remove it but remember we should use the ones you suggested next time.

**Comment by dtaniwaki** [comment]
@alexmt I tweaked the code and added default HDFS configuration support. Would you review it again? I'll squash the commits if necessary.

**Comment by alexmt** [comment]
Looks great!  Thank you @dtaniwaki !

**Comment by dtaniwaki** [comment]
@alexmt I made a separated PR of https://github.com/argoproj/argo/pull/1167 for more validations, which I found it is necessary while I'm implementing the HDFS artifact.

**Pull request merged**